ise.py

```diff
@@ -169,6 +171,22 @@ def _request(self, url, method="get", data=None):
 
         return req
 
+    def get_version(self):
+        try:
+            # Build MnT API URL
+            url =  "https://" + self.ise_node + "/admin/API/mnt/Version"
+            # Access MnT API
+            req = self.ise.request('get', url, data=None, timeout=self.timeout)
+            # Extract version of first node
+            soup = BeautifulSoup(req.content,'xml')
+            full_version = soup.find_all('version')[0].get_text()
+            # Get simple version ie: 2.7
+            short_version = float(full_version[0:3])
+            # print("ISE Initializing - Version Check " + full_version)
+            return short_version
+        except:
```

Comment:
Really not a fan of this broad exception. If there is a specific exception expected due to ISE version, that specifically should be caught.

Reply:
Keeping it simple - throwing an exception would break existing compatibility, as this PR maintains. As proposed, the library init code in this PR just continues as it does today for all users of the library. If bad data or credentials were entered, today the library only fails deeper in the code - on actual querying of data. This causes a library user to put more guardrails further down in the code and catch 403s when they were expecting API results. This change provides a simple way to check credentials, and if the Monitoring API was reachable, and fill in the version if available (it's a sanity check before calling an operation against the ERS REST API).  

Also note that ISE ERS Rest API in a multi node cluster is read-only if the Admin node is Standby. It's best to query ISE Monitoring API, to find which Admin node has the active role. If the library makes PUT request to a standby node, it will fail. This PR lays groundwork to get the active admin node from the MnT API, and makes the library more friendly to use by saving users from tripping and then trying to code around ISE API pitfalls like this. The only cost is a simple credential validation and query against the Monitoring API, before actually using the ERS REST API. I'll code that discovery in the next PR.

As an added bonus to confirming API credentials early, (later) obtaining the Active node, you get to discover the API version in the transaction. Knowing the version at class init would provide a way to block invalid API requests from being sent (sending a 3.x API request to a 2.x server, and vice versa). Or if a parameter changes between versions like we saw in #148, it provides a way to adapt the library to the API versions as they are released. I think the library has grown to a point that it needs a way to handle API version changes.

I think the concern you have is a small initialization time cost, but there are many benefits list above, when users will be counting on the accuracy and resiliency of this library rather than shaving off a few ms to turn off safety checks. If it's changed to be optional, I think it would be better on by default, and could be disabled if a user wants all the safeguards off.

Reply:
> I think the library has grown to a point that it needs a way to handle API version changes.

I disagree on this one. I’ve never seen an SDK that fetches things behind the scenes to determine what API version you want to reach.  A setting that we can pin, is what I expect and what I would want. For those in a situation where ISE gets patched without them knowing or testing against the patch, that sounds like a whole different issue. 

I understand the intent here, I just don’t think it should do all this without opting in. Adding “a small initialization time cost” to an SDK makes it less attractive for those who builds APIs using it. I understand scripts and scheduled tasks could benefit from this, but this is a downgrade in behavior for me.

If this is wanted in the library, I suggest it’s added as an optional setting. If it’s on by default I don’t really care, though I would argue I’d major bump it to give everyone a heads up.   
I don’t want my code (against a security system, especially) to be able to change dynamically without me knowing about it(or specifically asking for it). I’d rather have it fail. 


With all this said, you should the very least catch the error and log it as an exception.

Reply:
All in all, I think both of us will be happy with a setting for this. Whether it’s on by default or not don’t really matter to me personally, that’s something Falk probably should think about. 😊

Reply:
+1 that this should be optional. I don't have a problem with it being on by default, but please give me a way to opt out of the automatic version check.